### ISSUE TITLE
Using NodeJS or ASP.NET Core app

### ISSUE THREAD

**Issue by opcodewriter** [chromelyapps/Chromely#210]
I'm thinking of using Chromely to build a desktop app which is uses a server (NodeJS or an ASP.NET Core app, I haven't decided yet) and a client app written with HTML, JS, CSS.

Obviously, the server will need to listen to a local port and client will have to connect to it.

My concern is, because of using a localhost port, wouldn't be a problem with deploying my app? 
Wouldn't the users of my app get notified by the firewall about allowing the connection to work? This would be very ugly. And I think with https it gets even more complicated, because a certificate would be needed AFAIK...

I'd appreciate some input...

**Comment by Donaut** [comment]
Chromely can communicate with JavaScript. So you can call c# functions in JavaScript. I don't get it why you want to use Nodejs or Asp.net and use chromely to only render the Html page.

**Comment by opcodewriter** [comment]
@Donaut Thanks for your reply

> You clearly didn't used Chromely.

Sure, I didn't say I did.

> So you can call c# functions in JavaScript code.
> This way you don't need to use nodejs or asp.net at all.

I do want to use Node JS. Why? Because I'm using libraries which only NodeJS have.
Sorry but it doesn't answer my question.

**Comment by mattkol** [comment]
@opcodewriter 
In addition to @Donaut answer you can also check out the [demos](https://github.com/chromelyapps/demo-projects). I am sure one or a combination of them will help with your decisions. 

> I do want to use Node JS. Why? Because I'm using libraries which only NodeJS have.

We also use NodeJS because of the libs - see [angular-react-vue](https://github.com/chromelyapps/demo-projects/tree/master/angular-react-vue). You can build separately using [node/related cli](https://github.com/chromelyapps/demo-projects/blob/master/angular-react-vue/ChromelyAngular/angularapp/package.json). Or [include in project file build actions](https://github.com/chromelyapps/demo-projects/blob/cbb0eb2e930285d28813bbb3e60cc110de68ea0a/angular-react-vue/ChromelyAngular/ChromelyAngular.csproj#L59), [also see.](https://github.com/chromelyapps/demo-projects/blob/cbb0eb2e930285d28813bbb3e60cc110de68ea0a/angular-react-vue/ChromelyAngular/ChromelyAngular.csproj#L66)

> Are my concerns valid regarding the local web communication?

May be related to this - https://github.com/chromelyapps/Chromely/issues/201

From my experience so far, if you can browse to a URL locally using your default browser without Chromely, you should be able to do that with Chromely. Everything being equal of course. But you can see what we did [with Blazor](https://github.com/chromelyapps/demo-projects/tree/master/blazor) and [web-mvc](https://github.com/chromelyapps/demo-projects/tree/master/web-chromely) to see if it fits what you are looking for.

You did not say if this will be locally hosted or externally hosted for production. Unless there is need for it, like a mix of Web/Desktop or accessing a local site used by other applications too, I do not see the benefit of hosting a site locally and then accessing it using a framework like Chromely. If the local site is strictly for Chromely. It defeats the purpose. But since I do not understand the entire requirements, I won't say that is the case in this situation.

Thanks.

**Comment by opcodewriter** [comment]
Thank you @mattkol 

Some background: I started the work on a app which is a Node JS **server** app. It access files on disk and it's using some Node JS libraries (for example, creating XML on the fly).

> May be related to this - #201
Yes, I think that's exactly what I'm talking about!

>  But you can see what we did with Blazor and web-mvc to see if it fits what you are looking for.
Yes, I saw the samples, but like how you saw in #201, it can be a problem. 

> , I do not see the benefit of hosting a site locally and then accessing it using a framework like Chromely
I think the scenario is when you need to access disk or use server side libraries which need to do more than what it can be done as a client, either its JS or even Web assembly (through Web Assembly).

Please correct me if I am wrong, but I don't think Chromely can do this now.
I think Chromely needs at least two things: 
1. Start a "server" executable (not sure, I think it can already do this). A "server" is basically any app. It can be Node JS, an ASP.NET Core app, etc.
2. Have some inter-process communication with the "server". This is to overcome the issue with opening ports for web communication.
So basically Chromely needs an API which allows communicating between the app inside Chromely and the "server". 

Or maybe I am just missing something?

**Comment by mattkol** [comment]
@opcodewriter so let me see how I can help. First with explaining Chromely.

Chromely is primarily a desktop framework using html/csss/javascript. No server required. This is where it differs from Electron or Electron.NET. No server, no NodeJS or ASP.NET MVC or Core. None of that needed to use Chromely.

That is the fundamental. But it can be used in conjunction with a server. In that situation it is like any other browser. So it is redundant if you asked me. Now there are scenarios where developers use it to "embed remote content" in a mix-use scenario. There have also been situation where it is used for testing a live web application.

So let us break it down into 3, and you pick what scenario you are researching on so that we are more guided on how to help:

- 1. A pure desktop application where your create html pages of your own.
- 2. You access an external Web application.
- 3. Or a mix of i and ii

If this is a pure desktop application and according to you:
>  I started the work on a app which is a Node JS server app (I could rewrite it in .NET).

Then I will say stay with i. You can use the [Controllers](https://github.com/chromelyapps/demo-projects/tree/master/regular-chromely/CrossPlatDemo/Controllers) to write your .NET logic.

Try  the [base demo](https://github.com/chromelyapps/demo-projects/tree/master/regular-chromely/CrossPlatDemo) for an hour or so. That may answer a lot of questions. 

If this is for Windows, you can see an example on how to access files on Windows -https://github.com/chromelyapps/Chromely/issues/164#issuecomment-592953620
